### PR TITLE
ds 3d docs/llama2 example

### PR DESCRIPTION
**Pull request by ZhongYingMatrix**
### What changes were proposed in this pull request?

Please describe the changes you have made or proposed in this pull request.

### Why are the changes needed?

Explain the purpose or motivation behind these changes. What problem are you trying to solve?

### Does this PR introduce any user-facing change?

Specify whether this pull request introduces any changes that users will directly interact with or notice.

### How was this patch tested?

Detail the testing process you have undertaken to ensure the changes in this pull request are valid and working as intended.

### CODE REVIEW

File: atorch/examples/ds_3d_parallel/llama2_entry.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    "wall_clock_breakdown": false,
+    "pipeline": {
+      "activation_checkpoint_interval": 1
+    }
+  }
+EOT
+
+
+nohup python -u -m atorch.distributed.launch --nproc_per_node 8 --nnodes $WORLD_SIZE \
```
Comment (workingloong) [inline atorch/examples/ds_3d_parallel/llama2_entry.sh:52]: Use ` atorch.distributed.run` ？

File: atorch/examples/ds_3d_parallel/design.md [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+        )
+
+    # DeepSpeed PipelineEngine model
+    model = result.model
+    ```
+    </details>
+
+
+- Omits optim and dataset related batch_fn.
+- The user specifies the module name information of Tensor Paralle Shard, and the customized forward patcheres if some modules in the pipeline need.
```
Comment (workingloong) [inline atorch/examples/ds_3d_parallel/design.md:144]: "Paralle" -> "Parallel"

File: atorch/examples/ds_3d_parallel/design.md [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+# Design
+
+- The main problems with 3D parallelization of the native Torch implementation model:
+   - Graphics memory limitations, large-scale models can`t be instantiated as a whole on a single GPU before slicing. Also efficiency issues.
```
Comment (workingloong) [inline atorch/examples/ds_3d_parallel/design.md:148]: "can`t" -> "can't"

File: atorch/examples/ds_3d_parallel/design.md [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+   - Randomness control, model initialization/dropout, etc., need to control the same/different in different parallel dimensions.
+
+## Record init
+
+- Megatron`s embedded Col/Row Linear has been split horizontally and vertically, and deepspeed LayerSpec delays instantiating the required layers. But Transformers `XXXModel(XXXConfig(…))` instantiates the full model directly.
```
Comment (workingloong) [inline atorch/examples/ds_3d_parallel/design.md:153]: "Megatron`s" -> "Megatron's"

File: atorch/examples/ds_3d_parallel/design.md [hunk 5]
```diff
@@ -0,0 +1,844 @@
+# Goal
+- Large-scale model (100B+ levels) training, in addition to the FSDP zero3 scheme, the community more uses hybrid parallel (data parallel, pipeline parallel, tensor parallel, 3D) scheme. The most widely used frameworks include Megatron-LM (tensor parallelism), deepspeed (pipeline parallelism), that form hybrid parallelism with deep optimization.
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/design.md:2]: For large-scale model training (with 100B+ levels), besides using FSDP/zero3 parallelism, 3D parallelism is widely used in deep learning community. 3D parallelism includes tensor parallel, pipeline parallel, and  data parallel.  Megatron-LM and DeepSpeed provide excellent 3D parallelism implementation which are popular among users. 
 
Megatron-LM offers a Col/Row parallel layer that users can incorporate into the model definition to achieve tensor parallelism. DeepSpeed's pipeline parallel feature requires converting the model into a sequential list of LayerSpec, making its usage complicated especially for non-sequential models.

File: atorch/examples/ds_3d_parallel/design.md [hunk 6]
```diff
@@ -0,0 +1,844 @@
...
+                            forward_fn=_logits_helper,
+                            tied_weight_attr='word_embeddings_weight')
+            )
+    ```
+
+    </details>
+
+
+
+- The HuggingFace Transformers library provides native torch implementations of a large number of NLP models, which are constantly emerging. Megatron Deepspeed hybrid parallel support is limited.
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/design.md:99]: Don't need to mention Transformers.

File: atorch/examples/ds_3d_parallel/design.md [hunk 7]
```diff
@@ -0,0 +1,844 @@
...
+- The goal: to reuse deepspeed/megatron capabilities to provide a hybrid parallelism suitable for a wide range of native Torch implementations, and integrate into Atorch auto_accelerate optimization using a low-cost hybrid parallelism scheme.
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/design.md:100]: Just say:

ATorch supports 3D parallel training based on DeepSpeed/Megatron, and supports easy usage by using auto_accelerate with ds_3d_parallel optimization method.

File: atorch/examples/ds_3d_parallel/design.md [hunk 8]
```diff
@@ -0,0 +1,844 @@
...
+# User Interface
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/design.md:101]: Main purpose of this doc is explaining how to use ds_3d_parallel by example. 
So, many of the contexts can be omitted below.

Comment (ZhongYingMatrix) [reply atorch/examples/ds_3d_parallel/design.md:101]: `tutorial.md` explains how to use ds_3d_parallel by example, `design.md` mainly describes the implementation.

Comment (ZhongYingMatrix) [reply atorch/examples/ds_3d_parallel/design.md]: [remove design.md; merge intro to tutorial.md](https://github.com/intelligent-machine-learning/dlrover/pull/783/commits/a5fdcca2b03f36fd21017f90b6313b1346187132)

File: atorch/examples/ds_3d_parallel/llama2_entry.sh [hunk 9]
```diff
@@ -0,0 +1,56 @@
+set -x
+
+DATASET_PATH=${DATASET_PATH:-/path/to/wikitext-2-raw-v1}
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/llama2_entry.sh:3]: Any script that can download dataset/model if users don't have it ?

File: atorch/examples/ds_3d_parallel/ds_3d_llama2.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        model.train_batch_size(),
+        args.block_size,
+        llama_config.hidden_size,
+        llama_config.vocab_size,
+        llama_config.intermediate_size,
+        llama_config.num_hidden_layers,
+        use_gradient_checkpointing=model.module.activation_checkpoint_interval > 0,
+    )
+    timestamp = sync_and_time()
+    for iter in range(100):
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/ds_3d_llama2.py:155]: 100 is max_step?
then define a max_step args.

File: atorch/examples/ds_3d_parallel/ds_3d_llama2.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    def batch_fn(data):
+        tokens = data["input_ids"]
+        position_ids = data["position_ids"]
+        attention_mask = data["attention_mask"]
+        labels = data["labels"]
+        loss_mask = data["loss_mask"]
+
+        return (tokens, attention_mask, position_ids), (labels, loss_mask)
+
+    ds_config = args.ds_config
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/ds_3d_llama2.py:76]: Is this args.ds_config required? Or just leave it as default None?

File: atorch/examples/ds_3d_parallel/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,349 @@
+# Intro
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/tutorial.md:1]: Rename this file as README.md

File: atorch/examples/ds_3d_parallel/ds_3d_llama2.py (hunk 11 above)
Comment (ZhongYingMatrix) [reply atorch/examples/ds_3d_parallel/ds_3d_llama2.py:82]: https://github.com/intelligent-machine-learning/dlrover/blob/ds_3d_llama2/atorch/examples/ds_3d_parallel/llama2_entry.sh#L54

**Review by workingloong — APPROVED** [review]
LGTM

File: atorch/examples/ds_3d_parallel/llama2_entry.sh (hunk 9 above)
Comment (ZhongYingMatrix) [reply atorch/examples/ds_3d_parallel/llama2_entry.sh]: Done

File: atorch/examples/ds_3d_parallel/example_utils.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        raw_datasets = {}
+        try:
+            raw_datasets["train"] = load_dataset("wikitext", wiki_suffix, split="train")
+            raw_datasets["validation"] = load_dataset("wikitext", wiki_suffix, split="validation")
+            raw_datasets["test"] = load_dataset("wikitext", wiki_suffix, split="test")
+        except Exception as e:
+            logger.error(e)
+            from modelscope.msdatasets import MsDataset
+
+            raw_datasets["train"] = MsDataset.load("modelscope/wikitext", subset_name=wiki_suffix, split="train")
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/example_utils.py:57]: raw_datasets["train"] = raw_datasets["train"].to_torch_dataset()

File: atorch/examples/ds_3d_parallel/llama2_entry.sh [hunk 14]
```diff
@@ -0,0 +1,71 @@
+set -x
+
+# Dataset path, would download in `example_utils.py` if not exist
+DATASET_PATH=${DATASET_PATH:-/cache/wikitext-2-raw-v1}
+
+# Llama model path, download and convert it if not exist
+MODEL_SIZE=${MODEL_SIZE-7B}
+MODEL_NAME_OR_PATH=${MODEL_NAME_OR_PATH:-/cache/Llama-2-`echo $MODEL_SIZE|tr '[:upper:]' '[:lower:]'`-hf}
+if [ ! -d $MODEL_NAME_OR_PATH ]; then
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/llama2_entry.sh:9]: directory exists does not mean model file exists.

File: atorch/examples/ds_3d_parallel/llama2_entry.sh [hunk 15]
```diff
@@ -0,0 +1,71 @@
+set -x
+
+# Dataset path, would download in `example_utils.py` if not exist
+DATASET_PATH=${DATASET_PATH:-/cache/wikitext-2-raw-v1}
+
+# Llama model path, download and convert it if not exist
+MODEL_SIZE=${MODEL_SIZE-7B}
+MODEL_NAME_OR_PATH=${MODEL_NAME_OR_PATH:-/cache/Llama-2-`echo $MODEL_SIZE|tr '[:upper:]' '[:lower:]'`-hf}
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/llama2_entry.sh:8]: ~/.cache/....

File: atorch/examples/ds_3d_parallel/README.md [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+        </details>
+
+        
+
+      - `tie_first`: whether to tie the first module in deepspeed pipeline (e.g. vocab embedding in transformer). 
+      - `logit_helper`: take effect when `tie_first` is True. The helper function to compute logti on the tied embedding. `_default_logit_helper` will be used if None.
+      - `ds_config`: `Dict` for deepspeed config or `str` for json file path. 
+      - `batch_fn`:  input adaptation function for deepspeed pipeline (inputs,), (labels,). 
+
+#Example
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/README.md:340]: Use ## for sub-title.
Also ## Example.  (add black space)

File: atorch/examples/ds_3d_parallel/requirements.txt [hunk 17]
```diff
@@ -0,0 +1 @@
+datasets>=2.14.6
```
Comment (skydoorkai) [inline atorch/examples/ds_3d_parallel/requirements.txt:1]: + modelscope for dataset downloading.
\

Comment (ZhongYingMatrix) [reply atorch/examples/ds_3d_parallel/requirements.txt]: logger for install

**Review by skydoorkai — APPROVED** [review]
lgtm
